ls1intum/Artemis
# Bugfix/Deleting a participation with submissions + results

### Checklist

### Motivation and Context



### Description


Regression from https://github.com/ls1intum/Artemis/pull/695.
If a programming exercise participation is deleted that has a submission with a result, the deletion fails with this error:
```
Caused by: java.sql.SQLIntegrityConstraintViolationException: Cannot delete or update a parent row: a foreign key constraint fails (`ArTEMiS`.`result`, CONSTRAINT `FK3vct9sad5oubthdmq63n58mnp` FOREIGN KEY (`submission_id`) REFERENCES `submission` (`id`) ON DELETE NO ACTION ON UPDATE NO ACTION)
```

So the issue is that Hibernate tries to delete the submission before the result, which causes the foreign key constraint.

I made 3 small changes:
- Eagerly load the results and submissions in the delete service method (otherwise they appearently can't be loaded correctly in `deleteResultsAndSubmissionsOfParticipation`.
- Moved the logic for deleting submissions without a result into `deleteResultsAndSubmissionsOfParticipation`.
- Return the cleaned up participation without the submissions and results; otherwise Hibernate seems to try to delete the connected entities again when the participation is deleted. I think the issue is that there a 2 transactional contexts on top of each other here (1: delete, 2: deleteResultsAndSubmissionsOfParticipation)

I also wrote some integration tests & updated the docs.

:information_source: It is unclear to me why this isn't an issue for e.g. quizzes.

### Steps for Testing


1. Log in to ArTEMiS
2. Course administration > programming exercise > edit in editor
3. Create assignment participation
4. Submit for the assignment participation, wait for the result to return
5. Reset assignment repository (this first deletes the participation and then creates a new one)
-> There should be no error shown

### Screenshots
